File: frontend/weathergpt_app/lib/models/climate.dart
/// WeatherGPT — Climate Models (Phase 7)
/// Dart representations of deterministic climate trends, comparisons, and insights.
/// Aligned with backend/weathergpt_api/app/schemas/climate.py.
library;

/// A single year's aggregated value in a climate trend series.
class ClimateTrendPoint {
  const ClimateTrendPoint({
    required this.year,
    required this.value,
  });

  final int year;
  final double value;

  factory ClimateTrendPoint.fromJson(Map<String, dynamic> json) {
    return ClimateTrendPoint(
      year: (json['year'] as num).toInt(),
      value: (json['value'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
        'year': year,
        'value': value,
      };
}

/// Aggregated climate trend for a specific metric over a period.
class ClimateTrend {
  const ClimateTrend({
    required this.location,
    required this.metric,
    required this.unit,
    required this.period,
    required this.values,
  });

  final String location;
  final String metric;
  final String unit;
  final String period;
  final List<ClimateTrendPoint> values;

  factory ClimateTrend.fromJson(Map<String, dynamic> json) {
    return ClimateTrend(
      location: json['location'] as String? ?? '',
      metric: json['metric'] as String? ?? '',
      unit: json['unit'] as String? ?? '',
      period: json['period'] as String? ?? '',
      values: (json['values'] as List<dynamic>?)
              ?.map((e) => ClimateTrendPoint.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
        'location': location,
        'metric': metric,
        'unit': unit,
        'period': period,
        'values': values.map((e) => e.toJson()).toList(),
      };
}

/// Comparison of a recent period against a long-term historical baseline.
class ClimateComparison {
  const ClimateComparison({
    required this.metric,
    required this.currentValue,
    required this.historicalAverage,
    required this.difference,
    required this.differencePercent,
    required this.interpretation,
  });

  final String metric;
  final double currentValue;
  final double historicalAverage;
  final double difference;
  final double differencePercent;
  final String interpretation;

  bool get isAboveAverage => interpretation == 'above_average';
  bool get isBelowAverage => interpretation == 'below_average';
  bool get isNearAverage => interpretation == 'near_average';

  factory ClimateComparison.fromJson(Map<String, dynamic> json) {
    return ClimateComparison(
      metric: json['metric'] as String? ?? '',
      currentValue: (json['current_value'] as num?)?.toDouble() ?? 0.0,
      historicalAverage: (json['historical_average'] as num?)?.toDouble() ?? 0.0,
      difference: (json['difference'] as num?)?.toDouble() ?? 0.0,
      differencePercent: (json['difference_percent'] as num?)?.toDouble() ?? 0.0,
      interpretation: json['interpretation'] as String? ?? 'near_average',
    );
  }

  Map<String, dynamic> toJson() => {
        'metric': metric,
        'current_value': currentValue,
        'historical_average': historicalAverage,
        'difference': difference,
        'difference_percent': differencePercent,
        'interpretation': interpretation,
      };
}

/// Full climate analysis response returned by GET /api/v1/climate.
class ClimateResponse {
  const ClimateResponse({
    required this.location,
    required this.yearFrom,
    required this.yearTo,
    required this.temperatureTrend,
    required this.rainfallTrend,
    required this.temperatureComparison,
    required this.rainfallComparison,
    required this.temperatureAnomaly,
    required this.rainfallAnomaly,
    required this.season,
    required this.insight,
    required this.dataSource,
    this.availableLocations = const [],
  });

  final String location;
  final int yearFrom;
  final int yearTo;
  final ClimateTrend temperatureTrend;
  final ClimateTrend rainfallTrend;
  final ClimateComparison temperatureComparison;
  final ClimateComparison rainfallComparison;
  final double temperatureAnomaly;
  final double rainfallAnomaly;
  final String season;
  final String insight;
  final String dataSource;
  final List<String> availableLocations;

  factory ClimateResponse.fromJson(Map<String, dynamic> json) {
    return ClimateResponse(
      location: json['location'] as String? ?? '',
      yearFrom: (json['year_from'] as num?)?.toInt() ?? 2000,
      yearTo: (json['year_to'] as num?)?.toInt() ?? 2023,
      temperatureTrend: ClimateTrend.fromJson(
        json['temperature_trend'] as Map<String, dynamic>? ?? {},
      ),
      rainfallTrend: ClimateTrend.fromJson(
        json['rainfall_trend'] as Map<String, dynamic>? ?? {},
      ),
      temperatureComparison: ClimateComparison.fromJson(
        json['temperature_comparison'] as Map<String, dynamic>? ?? {},
      ),
      rainfallComparison: ClimateComparison.fromJson(
        json['rainfall_comparison'] as Map<String, dynamic>? ?? {},
      ),
      temperatureAnomaly: (json['temperature_anomaly'] as num?)?.toDouble() ?? 0.0,
      rainfallAnomaly: (json['rainfall_anomaly'] as num?)?.toDouble() ?? 0.0,
      season: json['season'] as String? ?? '',
      insight: json['insight'] as String? ?? '',
      dataSource: json['data_source'] as String? ?? '',
      availableLocations: (json['available_locations'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
        'location': location,
        'year_from': yearFrom,
        'year_to': yearTo,
        'temperature_trend': temperatureTrend.toJson(),
        'rainfall_trend': rainfallTrend.toJson(),
        'temperature_comparison': temperatureComparison.toJson(),
        'rainfall_comparison': rainfallComparison.toJson(),
        'temperature_anomaly': temperatureAnomaly,
        'rainfall_anomaly': rainfallAnomaly,
        'season': season,
        'insight': insight,
        'data_source': dataSource,
        'available_locations': availableLocations,
      };
}

File: backend/weathergpt_api/app/schemas/climate.py
"""
WeatherGPT — Climate Schemas (Phase 7)
Pydantic models for historical climate analysis request/response validation.

All calculations are deterministic and data-driven (no LLM involvement).
"""

from pydantic import BaseModel, Field
from typing import List, Optional


# ---------------------------------------------------------------------------
# Records — raw dataset row representation
# ---------------------------------------------------------------------------

class HistoricalClimateRecord(BaseModel):
    """A single monthly climate record from the historical dataset."""

    location: str = Field(..., description="City name")
    year: int = Field(..., description="Year of the record")
    month: int = Field(..., ge=1, le=12, description="Month (1–12)")
    avg_temp_c: float = Field(..., description="Average temperature (°C)")
    max_temp_c: float = Field(..., description="Maximum temperature (°C)")
    min_temp_c: float = Field(..., description="Minimum temperature (°C)")
    rainfall_mm: float = Field(..., description="Total rainfall (mm)")


# ---------------------------------------------------------------------------
# Trend — yearly or monthly aggregated series
# ---------------------------------------------------------------------------

class ClimateTrendPoint(BaseModel):
    """A single year's aggregated value in a trend series."""

    year: int = Field(..., description="Year")
    value: float = Field(..., description="Aggregated metric value")


class ClimateTrend(BaseModel):
    """Aggregated climate trend for a specific metric over a period."""

    location: str = Field(..., description="City name")
    metric: str = Field(..., description="'temperature' or 'rainfall'")
    unit: str = Field(..., description="Unit of measurement (e.g. '°C', 'mm')")
    period: str = Field(..., description="Human-readable period e.g. '2000–2023'")
    values: List[ClimateTrendPoint] = Field(
        default_factory=list, description="Year-by-year aggregated values"
    )


# ---------------------------------------------------------------------------
# Comparison — current period vs historical baseline
# ---------------------------------------------------------------------------

class ClimateComparison(BaseModel):
    """Comparison of a recent period against the long-term historical baseline."""

    metric: str = Field(..., description="'temperature' or 'rainfall'")
    current_value: float = Field(..., description="Recent period average/total")
    historical_average: float = Field(..., description="Long-term baseline average/total")
    difference: float = Field(..., description="current_value − historical_average")
    difference_percent: float = Field(
        ..., description="Percentage difference relative to baseline"
    )
    interpretation: str = Field(
        ..., description="'above_average' | 'near_average' | 'below_average'"
    )


# ---------------------------------------------------------------------------
# Full API response
# ---------------------------------------------------------------------------

class ClimateResponse(BaseModel):
    """
    Full climate analysis response returned by GET /api/v1/climate.

    Contains temperature and rainfall trends, historical comparisons,
    anomalies, seasonal context, and a rule-based textual insight.
    """

    location: str = Field(..., description="Requested city name")
    year_from: int = Field(..., description="Start year of the analysis")
    year_to: int = Field(..., description="End year of the analysis")

    temperature_trend: ClimateTrend
    rainfall_trend: ClimateTrend

    temperature_comparison: ClimateComparison
    rainfall_comparison: ClimateComparison

    temperature_anomaly: float = Field(
        ..., description="Temperature deviation from the full-period baseline (°C)"
    )
    rainfall_anomaly: float = Field(
        ..., description="Rainfall deviation from the full-period baseline (mm)"
    )

    season: str = Field(
        ...,
        description=(
            "Dominant Tamil Nadu season for the queried period: "
            "'Winter' | 'Summer' | 'Southwest Monsoon' | 'Northeast Monsoon'"
        ),
    )
    insight: str = Field(
        ..., description="Rule-based textual climate insight (no LLM)"
    )
    data_source: str = Field(
        default=(
            "Prototype/reference dataset — derived from publicly available "
            "climatological summaries for Tamil Nadu cities. "
            "Not official meteorological observations."
        ),
        description="Data provenance notice",
    )

    available_locations: List[str] = Field(
        default_factory=list,
        description="All locations available in the dataset",
    )


# ---------------------------------------------------------------------------
# Query schema (for documentation purposes; FastAPI reads query params directly)
# ---------------------------------------------------------------------------

class ClimateQueryParams(BaseModel):
    """Query parameters accepted by the climate endpoint."""

    location: str = Field("Madurai", description="City name")
    year_from: int = Field(2000, ge=1990, le=2030, description="Start year")
    year_to: int = Field(2023, ge=1990, le=2030, description="End year")
    month: Optional[int] = Field(
        None, ge=1, le=12, description="Filter to a single month (1–12); omit for full year"
    )
